dbms/src/Server/Server.cpp

```diff
@@ -1250,9 +1247,51 @@ try
             LOG_ERROR(log, "CPUAffinityManager::bindThreadCPUAffinity throws exception.");
         }
 
+        // Ready to provide service
+        tmt_context.setStatusRunning();
+
         LOG_INFO(log, "Start to wait for terminal signal");
         waitForTerminationRequest();
 
+        LOG_INFO(log, "Set store context status Stopping");
+        tmt_context.setStatusStopping();
+
+        // Wait for all mpp tasks to finish
+        auto monitor = tmt_context.getMPPTaskManager()->getMPPTaskMonitor();
+        auto start = std::chrono::steady_clock::now();
+        // The maximum seconds TiFlash will wait for all current MPP tasks to finish before shutting down
+        static constexpr const char * GRACEFUL_WIAT_BEFORE_SHUTDOWN = "flash.graceful_wait_before_shutdown";
+        auto graceful_wait_before_shutdown
+            = global_context->getUsersConfig()->getUInt64(GRACEFUL_WIAT_BEFORE_SHUTDOWN, 60);
+        auto max_wait_time = start + std::chrono::seconds(graceful_wait_before_shutdown);
+        // Sleep before checking to reduce the chance of missing MPP tasks that are still in the process of being dispatched
+        std::this_thread::sleep_for(std::chrono::milliseconds(200));
+        while (true)
+        {
+            {
+                std::unique_lock lock(monitor->mu);
+                if (monitor->monitored_tasks.empty())
+                    break;
+            }
+            auto current_time = std::chrono::steady_clock::now();
+            if (current_time >= max_wait_time)
+            {
+                LOG_WARNING(log, "Timed out waiting for MPP tasks to finish after {}s", graceful_wait_before_shutdown);
+                break;
+            }
+
+            std::this_thread::sleep_for(std::chrono::milliseconds(200));
+        }
+
+        proxy_machine.waitAllReadIndexTasksFinish(tmt_context);
+
+        LOG_INFO(log, "Set store context status Terminated");
+        tmt_context.setStatusTerminated();
+
+        // Stop grpc server
+        flash_grpc_server_holder.reset();
+        proxy_machine.stopProxy(tmt_context);
```

Comment:
The `LocalAdmissionController` and `TaskScheduler` is also reset before proxy is stopped in the original code. Do we need to move them? I'm not familiar with those two structs.

Reply:
I checked the code. These two structures have nothing to do with the proxy.